fix(auth): Fix several typing inconsistencies

MathBunny opened the pull request:
While working on the modularization of `instance-id` I came across several typing inconsistencies that I thought I'd bring up. Not sure if this is the best approach for all of them though.

##### 1) `auth-config.ts`
Observation: The `enabled` field should not be optional in `auth-config.ts`.

Source
https://github.com/firebase/firebase-admin-node/blob/54a58348ab7d8ee5e1ef4fd90cb68110a3e41391/src/auth/auth-config.ts#L168-L170

Typing

https://github.com/firebase/firebase-admin-node/blob/54a58348ab7d8ee5e1ef4fd90cb68110a3e41391/src/auth.d.ts#L960-L972

##### 2) `auth.ts`
Observation: There is a missing `uid` field in `DecodedIdToken`.

Source
https://github.com/firebase/firebase-admin-node/blob/54a58348ab7d8ee5e1ef4fd90cb68110a3e41391/src/auth/auth.ts#L108-L111

Typing
https://github.com/firebase/firebase-admin-node/blob/54a58348ab7d8ee5e1ef4fd90cb68110a3e41391/src/auth.d.ts#L531-L538

##### 3) `user-record.ts`
Observation: `displayName` in `MultiFactorInfo` doesn't have a `null` option for type.

Source
https://github.com/firebase/firebase-admin-node/blob/54a58348ab7d8ee5e1ef4fd90cb68110a3e41391/src/auth/user-record.ts#L149-L152

Typing
https://github.com/firebase/firebase-admin-node/blob/54a58348ab7d8ee5e1ef4fd90cb68110a3e41391/src/auth.d.ts#L80-L90

##### 4) `user-record.ts`
Observation: `enrolledFactors ` in `MultiFactor ` isn't read-only in types.

Source
https://github.com/firebase/firebase-admin-node/blob/54a58348ab7d8ee5e1ef4fd90cb68110a3e41391/src/auth/user-record.ts#L267-L268

Typing
https://github.com/firebase/firebase-admin-node/blob/54a58348ab7d8ee5e1ef4fd90cb68110a3e41391/src/auth.d.ts#L228-L233

hiranya911 commented:
Haven't dug into the details (I'll let @bojeil-google make a more informed assessment). My gut feeling is:

* 1-2 are trivial changes that we can do right away.
* 3 possibly requires updating the d.ts file to make `displayName` nullable instead of optional.
* 4 we should probably change the implementation to use a regular array. I don't know if we want to expose `ReadonlyArray` in our public typings at the moment.

MathBunny commented:
Thanks @hiranya911. Regarding the fourth point, indeed my approach right now is to change the implementation to use a regular array.

bojeil-google commented on test/unit/auth/auth.spec.ts:
```diff
@@ -161,6 +161,7 @@ function getDecodedIdToken(uid: string, authTime: Date, tenantId?: string): Deco
       sign_in_provider: 'custom', // eslint-disable-line @typescript-eslint/camelcase
       tenant: tenantId,
     },
+    uid: '123',
```
This should match the `sub` field, eg. the value should be `uid`

bojeil-google commented on test/unit/auth/auth.spec.ts:
```diff
@@ -186,6 +187,7 @@ function getDecodedSessionCookie(uid: string, authTime: Date, tenantId?: string)
       sign_in_provider: 'custom', // eslint-disable-line @typescript-eslint/camelcase
       tenant: tenantId,
     },
+    uid: '123',
```
Same here. This should match the `sub` field, eg. the value should be `uid`

bojeil-google commented on src/auth/user-record.ts:
```diff
@@ -148,7 +148,7 @@ export enum MultiFactorId {
 export abstract class MultiFactorInfo {
   public readonly uid: string;
-  public readonly displayName: string | null;
+  public readonly displayName: string;
```
Hmm, actually this is nullable. Check below: 
`utils.addReadonlyGetter(this, 'displayName', response.displayName || null);`

If we can't fix in the type definitions, then we should modify below to not be nullable.

hiranya911 replied:
I think changing typings from optional to nullable can be a breaking change. I think fixing the implementation to match the public API declaration is the right move here.

MathBunny commented:
Thanks for the feedback @bojeil-google, I addressed your comments, PTAL :) 

Not sure about the unit tests and if the `displayName` change is a good idea.

bojeil-google commented:
@hiranya911 what are your thoughts on `displayName`? I personally prefer changing the d.ts to make this nullable, but we can also update the implementation as is currently done to not be nullable.

hiranya911 commented on src/auth/auth.ts:
```diff
@@ -108,6 +108,7 @@ export interface DecodedIdToken {
   sub: string;
   tenant?: string;
   [key: string]: any;
+  uid: string;
```
Move this up, so the `[key: string]` becomes the last entry.

MathBunny commented:
Just as a note, I'm actually a fan of making it nullable, I just always steer more towards not changing typings. If possible I'd go that route
